File: forms/sports-medicine-assessment/full-stack-with-loco-tera-htmx-alpine/src/engine/ppe_grader.rs
use super::flagged_issues::detect_additional_flags;
use super::ppe_rules::all_rules;
use super::types::{AssessmentData, Clearance, FiredRule, GradingResult};

/// Aggregate clearance from the firedRules list (highest grade wins).
///
///   any rule with grade 4  -> not-cleared
///   else any grade 3       -> pending
///   else any grade 2       -> conditional
///   else                   -> cleared
pub fn aggregate_clearance(fired_rules: &[FiredRule]) -> Clearance {
    let max_grade = fired_rules.iter().map(|r| r.grade).max().unwrap_or(0);
    match max_grade {
        g if g >= 4 => "not-cleared".to_string(),
        3 => "pending".to_string(),
        2 => "conditional".to_string(),
        _ => "cleared".to_string(),
    }
}

/// Evaluate every PPE rule against the data, returning the rules that fired
/// (sorted highest grade first).
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out: Vec<FiredRule> = Vec::new();
    for rule in all_rules() {
        if (rule.fires)(data) {
            out.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                grade: rule.grade,
            });
        }
    }
    // Sort highest grade first so the report reads worst-first.
    out.sort_by(|a, b| b.grade.cmp(&a.grade));
    out
}

/// Pure function: grades a Sports Medicine PPE assessment.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let fired_rules = run_rules(data);
    let additional_flags = detect_additional_flags(data);
    let clearance = aggregate_clearance(&fired_rules);
    let answered_count = super::utils::answered_count(data);

    GradingResult {
        clearance,
        answered_count,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

File: forms/sports-medicine-assessment/front-end-with-html/js/ppe-grader.js
import { ppeRules } from './rules.js';

// PPE clearance grader. Pure functions: take an `AssessmentData` object,
// run every rule in `ppeRules`, and aggregate the highest-grade rule that
// fired into a single `Clearance` decision.
//
// Aggregation:
//   any rule with grade 4  -> not-cleared
//   else any grade 3       -> pending
//   else any grade 2       -> conditional
//   else                   -> cleared

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').Clearance} Clearance
 * @typedef {import('./types.js').FiredRule} FiredRule
 */

/**
 * @param {FiredRule[]} firedRules
 * @returns {Clearance}
 */
function aggregateClearance(firedRules) {
  let maxGrade = 0;
  for (const r of firedRules) {
    if (r.grade > maxGrade) maxGrade = r.grade;
  }
  if (maxGrade >= 4) return 'not-cleared';
  if (maxGrade === 3) return 'pending';
  if (maxGrade === 2) return 'conditional';
  return 'cleared';
}

/**
 * Run every PPE rule against the supplied assessment data and produce
 * the resulting clearance decision.
 *
 * @param {AssessmentData} data
 * @returns {{ clearance: Clearance, firedRules: FiredRule[] }}
 */
function gradePPE(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];

  for (const rule of ppeRules) {
    try {
      if (rule.fires(data)) {
        firedRules.push({
          id: rule.id,
          category: rule.category,
          description: rule.description,
          grade: rule.grade
        });
      }
    } catch (e) {
      console.warn(`PPE rule ${rule.id} evaluation failed:`, e);
    }
  }

  // Sort highest grade first so the report reads worst-first.
  firedRules.sort((a, b) => b.grade - a.grade);

  const clearance = aggregateClearance(firedRules);
  return { clearance, firedRules };
}

export { gradePPE, aggregateClearance };
